taichi/program/kernel_profiler.cpp

```diff
@@ -131,129 +138,15 @@ class DefaultProfiler : public KernelProfilerBase {
-          constexpr float64 cuda_time_offset = 3e-4;
-          // Since event recording and synchronization can take 5 us, it's hard
-          // to exactly measure the real event time. Also note there seems to be
-          // a systematic time offset on CUDA, so adjust for that using a 3e-4 s
-          // magic number.
-          base_time_ = final_t + cuda_time_offset;
-        } else {
-          CUDADriver::get_instance().event_destroy(e);
-        }
-      }
-    }
-
-    return stop;
-#else
-    TI_NOT_IMPLEMENTED;
-#endif
-  }
-
-  virtual void stop(TaskHandle handle) override {
-#if defined(TI_WITH_CUDA)
-    CUDADriver::get_instance().event_record(handle, 0);
-#else
-    TI_NOT_IMPLEMENTED;
-#endif
-  }
-
-  std::string title() const override {
-    return "CUDA Profiler";
-  }
-
-  void sync() override {
-#if defined(TI_WITH_CUDA)
-    CUDADriver::get_instance().stream_synchronize(nullptr);
-    auto &timeline = Timeline::get_this_thread_instance();
-    for (auto &map_elem : outstanding_events) {
-      auto &list = map_elem.second;
-      for (auto &item : list) {
-        auto start = item.first, stop = item.second;
-        float kernel_time;
-        CUDADriver::get_instance().event_elapsed_time(&kernel_time, start,
-                                                      stop);
-
-        if (Timelines::get_instance().get_enabled()) {
-          float time_since_base;
-          CUDADriver::get_instance().event_elapsed_time(&time_since_base,
-                                                        base_event_, start);
-          timeline.insert_event({map_elem.first, true,
-                                 base_time_ + time_since_base * 1e-3, "cuda"});
-          timeline.insert_event(
-              {map_elem.first, false,
-               base_time_ + (time_since_base + kernel_time) * 1e-3, "cuda"});
-        }
-
-        auto it = std::find_if(
-            records.begin(), records.end(),
-            [&](KernelProfileRecord &r) { return r.name == map_elem.first; });
-        if (it == records.end()) {
-          records.emplace_back(map_elem.first);
-          it = std::prev(records.end());
-        }
-        it->insert_sample(kernel_time);
-        total_time_ms += kernel_time;
-
-        // TODO: the following two lines seem to increases profiler overhead a
-        // little bit. Is there a way to avoid the overhead while not creating
-        // too many events?
-        CUDADriver::get_instance().event_destroy(start);
-        CUDADriver::get_instance().event_destroy(stop);
-      }
-    }
-    outstanding_events.clear();
-#else
-    printf("CUDA Profiler not implemented;\n");
-#endif
-  }
-
-  static KernelProfilerCUDA &get_instance() {
-    static KernelProfilerCUDA profiler;
-    return profiler;
-  }
-
- private:
-  void *base_event_{nullptr};
-  float64 base_time_;
-};
 }  // namespace
 
 std::unique_ptr<KernelProfilerBase> make_profiler(Arch arch) {
   if (arch == Arch::cuda) {
+#if defined(TI_WITH_CUDA)
```

Comment:
Hmmm maybe this is related to my refactor, if you move `make_profiler` in `program.cpp` after the LlvmProgramImpl construction block, it might save you one `#if/else` here. 
(feel free to ignore this if that doesn't work ;)

Reply:
I’d like to try it when decoupling `profiler` from `ti.init()` and `program` ( about to begin ) , thanks !